adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala

```diff
@@ -2135,6 +2163,41 @@ trait MultisampleGenomicRDD[T, U <: MultisampleGenomicRDD[T, U]] extends Genomic
   }
 }
 
+/**
+ * A trait describing a GenomicDataset that also supports the Spark SQL APIs.
+ */
+trait DatasetBoundGenomicDataset[T, U <: Product, V <: GenomicDataset[T, U, V]] extends GenomicDataset[T, U, V] {
```

Comment:
@fnothaft I think you may have suggested this, doesn't `GenomicDataset` already imply the RDD has been bound to a dataset?  Couldn't these be moved there?

I'm not very sure what `GenomicDataset` is for, it isn't used everywhere I would've thought
https://github.com/bigdatagenomics/adam/search?utf8=%E2%9C%93&q=GenomicDataset&type=

Reply:
No, `GenomicDataset` is any `GenomicRDD` where the underlying genomic data can be viewed as both an RDD or a Dataset, while the base `GenomicRDD` trait only provides RDDs. This distinction will go away soon.

Reply:
`ParquetUnbound`/`RDDBound`/`DatasetBound` implies how the data is currently physically represented. Parquet unbound means that the data hasn't been materialized into memory yet and that the data is represented using a Parquet file, while RDD/DatasetBound means that the current (as of the last transformation) representation is an RDD/Dataset.

Since `cache` impacts the physical materialization of the data, we want the new `DatasetBoundGenomicDataset` abstract class to override `cache`, and not the `GenomicDataset` abstract class.